File: src/go/4_BehavioralPatterns/8_State/classic.go
package state

import "fmt"

type Switch struct {
  State State
}

func NewSwitch() *Switch {
  return &Switch{NewOffState()}
}

func (s *Switch) On() {
  s.State.On(s)
}

func (s *Switch) Off() {
  s.State.Off(s)
}

type State interface {
  On(sw *Switch)
  Off(sw *Switch)
}

type BaseState struct {}

func (s *BaseState) On(sw *Switch) {
  fmt.Println("Light is already on")
}

func (s *BaseState) Off(sw *Switch) {
  fmt.Println("Light is already off")
}

type OnState struct {
  BaseState
}

func NewOnState() *OnState {
  fmt.Println("Light turned on")
  return &OnState{BaseState{}}
}

func (o *OnState) Off(sw *Switch) {
  fmt.Println("Turning light off...")
  sw.State = NewOffState()
}

type OffState struct {
  BaseState
}

func NewOffState() *OffState {
  fmt.Println("Light turned off")
  return &OffState{BaseState{}}
}

func (o *OffState) On(sw *Switch) {
  fmt.Println("Turning light on...")
  sw.State = NewOnState()
}

func main() {
  sw := NewSwitch()
  sw.On()
  sw.Off()
  sw.Off()
}

File: src/js/4_BehavioralPatterns/8_State/classic.js
class Switch
{
  constructor()
  {
    this.state = new OffState();
  }

  on()
  {
    this.state.on(this);
  }

  off()
  {
    this.state.off(this);
  }
}

class State
{
  constructor()
  {
    if (this.constructor === State)
      throw new Error('abstract!');
  }

  on(sw)
  {
    console.log('Light is already on.');
  }

  off(sw)
  {
    console.log('Light is already off.')
  }
}

class OnState extends State
{
  constructor()
  {
    super();
    console.log('Light turned on.')
  }

  off(sw) {
    console.log('Turning light off...');
    sw.state = new OffState();
  }
}

class OffState extends State
{
  constructor()
  {
    super();
    console.log('Light turned off.');
  }

  on(sw) {
    console.log('Turning light on...');
    sw.state = new OnState();
  }
}

let sw = new Switch();
sw.on();
sw.off();
sw.off();